lib/recursive-methods.rb

```diff
@@ -1,49 +1,90 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) factorial will be called n number of times, and everything within the function is constant
+# Space complexity: O(n) n number of stack frames will be used
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 1 || n == 0
+    return 1
+  elsif n < 0
+    raise ArgumentError
+  end
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the length of s, reverse will get called n/2 times
```

Comment:
This is actually O(n^2) since `s[1...-1]` creates a new string of length n-2 each iteration.  This is true for both space & time.